libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -1620,6 +1623,30 @@ bool Compass::configured(char *failure_msg, uint8_t failure_msg_len)
     return all_configured;
 }
 
+void Compass::_reset_compass_id()
+{
+#if COMPASS_MAX_INSTANCES > 1
+    // Check if any of the registered devs are not registered
+    for (Priority i(0); i<COMPASS_MAX_INSTANCES; i++) {
+        if (_get_state(i).dev_id == 0) {
+            _priority_did_stored_list[i].set_and_save(0);
+        }
+    }
+    
+    // Check if any of the registered devs are not registered
+    for (StateIndex i(0); i<COMPASS_MAX_INSTANCES; i++) {
+        if (_state[i].dev_id == 0) {
+            // also hard reset dev_ids that are not detected
+            _state[i].dev_id.save();
+            gcs().send_text(MAV_SEVERITY_ALERT, "Mag: Compass #%d with DEVID %lu deregistered!", uint8_t(i), (unsigned long)_state[i].expected_dev_id);
```

Comment:
should not do this if _state[i].expected_dev_id is zero. I see:
APM: Mag: Compass #2 with DEVID 0 deregistered!

Reply:
Yeah, that's a bug. Fixed